### DISCUSSION TITLE
How to update the session environment on Windows runner?

### DISCUSSION THREAD

**Discussion by MarkCallow** [actions/virtual-environments#6065] [General questions] [answered]
I am migrating my Windows CI build to Actions. I need to install some software not available in Chocolatey, e.g. the Vulkan SDK. The installers for these items set some environment variables in the registry and update the Path environment variable. In my previous CI I used `refreshenv`, which is part of the Chocolatey installation, to pull the updated environment into the current session. This is not working on the Actions runner.  See this [example build](https://github.com/KhronosGroup/KTX-Software/runs/7761608976?check_suite_focus=true).

Here is an extract.

    - name: Install Dependencies
      run: |
        ci_scripts/install_win.bat
        echo "Pick up environment changes made to registry by the install script."
        refreshenv
        echo "VULKAN_SDK is $env:VULKAN_SDK"
        reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /v VULKAN_SDK

    - name: Build Windows
      run: |
        reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /v VULKAN_SDK
        echo "VULKAN_SDK is $env:VULKAN_SDK"
        ci_scripts/build_win.ps1

The _Install Dependencies_ step calls a script to download and run the Vulkan SDK installer. The installer sets the environment variable VULKAN_SDK into the Windows registry. After the script, I call Chocolatey's refreshenv to get the variable into the current session however it is not visible.  The reg query shows it is set in the registry.

It also does not appear in the _Build Windows_ step. Isn't this a new instance of Powershell? If so it so should get whatever was added to registry during _Install Dependencies_. Again the reg query shows the env. var. is present.

One strange thing is that the message `refreshenv` prints specifically mentions cmd.exe: `Refreshing environment variables from registry for cmd.exe. Please wait...Finished..` If somehow the cmd.exe version of `refreshenv` is being run it would explain why Powershell is not seeing the changes.

Another is that according to this [Chocolatey documentation](https://docs.chocolatey.org/en-us/create/functions/update-sessionenvironment) `refreshenv` is just an alias for `Update-SessionEnvironment` but if I replace `refreshenv` with `Update-SessionEnvironment` I get `The term 'Update-SessionEnvironment' is not recognized as a name of a cmdlet`

How can I get the _Build Windows_ step to see the environment variable?

**Comment by al-cheb** [comment]
Hello @MarkCallow.
You should use $GITHUB_ENV to set up a new env variable - https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files

```
 - run: |
          "VULKAN_SDK=something" >> $env:GITHUB_ENV
 - run: |
          $env:VULKAN_SDK
```

![image](https://user-images.githubusercontent.com/47745270/184306020-eb68dcae-cb7a-4360-b3e2-74d6ff04f118.png)

**Reply by MarkCallow** [reply]
Thank you. The problem with having to do it this way is that my yaml now needs to know what environment variables were set by the installers run by the `install_win`, whether any of those installers augment Path and if so, with what directory. If they do I have to add the Path additions to GITHUB_PATH. Complicating this is that what installers are run by the install script is determined by various project feature settings. If I have to set things to GITHUB_{ENV,PATH} in the yaml I'll have to test all those settings again.

It looks horrendously complicated compared to running a tool to update the environment from the registry.  When I install something with Chocolatey, which I am also doing, it just works. It seems that Chocolatey is managing to to run its refreshenv and make the updated environment available to later steps. How does it do it?

Since these variables  _are_ in the registry why does the _Build Windows_ step not pick them up. Is it really run in the same instance of Powershell as the _Install Dependencies_ step?

I build my project with CMake and there are CMake find modules to find the packages being installed. Therefore neither my CMakeLists nor the yaml file for my current CI have nor need any knowledge of the environment variables or Path additions made by various packages.

**Comment by MarkCallow** [comment, marked as answer]
I found the solution.  To use Chocolatey's `refreshenv`/`Update-SesionEnvironment` in Powershell (PS) you have to `import` their helpers package to your PS.

```
Import-Module "$env:ChocolateyInstall/helpers/chocolateyInstaller.psm1"
refreshenv
```

The Chocolatey install instructions that I found do not mention the helpers. I do not know if having to import them is standard operating procedure.

Although the runner creates a new PS for each step, those processes do not see the registry updates which means they must be launched from an existing PS instance and not the system. Therefore you need to add the above to any step that needs the environment variables.

If you are using Java you also have to be aware that the registry contains a `JAVA_HOME` env. var. which is almost certainly not the one you want. If, in a previous step you set up Java, e.g. using the `setup-java` action, then you must do the following:

```
$step_java_home=$env:JAVA_HOME
Import-Module "$env:ChocolateyInstall/helpers/chocolateyInstaller.psm1"
refreshenv
$env:JAVA_HOME=$step_java_home
```

I thought about creating something like `refreshenv` that would add new variables to $GITHUB_ENV and new $Path elements to $GITHUB_PATH but as I have only one step that needs the new variables, it wasn't worth the effort.

**Reply by gep13** [reply]
Using Import-Module is perfectly valid here.  By default, the `refreshenv` helper works within a Command Prompt, however, not within a PowerShell session. To get it to work within a PowerShell session, you need to do the Import-Module.  Typically, we recommend folks add this to their PowerShell Profile: https://docs.chocolatey.org/en-us/troubleshooting#why-does-choco-tab-not-work-for-me
